layers/mpls_echo.go

```diff
@@ -0,0 +1,550 @@
+		return fmt.Errorf("Invalid MPLSEcho content - length %v less than 32 bytes", len(data))
+	}
+
+	totalTLVBytes := len(data) - 32
+	if totalTLVBytes%4 != 0 {
+		return fmt.Errorf("Invalid MPLSEcho content - TLVs aren't 4-octet aligned (TLV bytes = %v)", totalTLVBytes)
+	}
+
+	m.VersionNumber = MPLSEchoVersion(binary.BigEndian.Uint16(data[0:2]))
+	m.GlobalFlags = MPLSEchoGlobalFlags(binary.BigEndian.Uint16(data[2:4]))
+	m.MessageType = MPLSEchoMessageType(data[4])
+	m.ReplyMode = MPLSEchoReplyMode(data[5])
+	m.ReturnCode = MPLSReplyReturnCode(data[6])
+	m.ReturnSubcode = data[7]
+	m.SenderHandle = binary.BigEndian.Uint32(data[8:12])
+	m.SequenceNumber = binary.BigEndian.Uint32(data[12:16])
+	m.TimestampSentSeconds = binary.BigEndian.Uint32(data[16:20])
+	m.TimestampSentSecondsFraction = binary.BigEndian.Uint32(data[20:24])
+	m.TimestampReceivedSeconds = binary.BigEndian.Uint32(data[24:28])
+	m.TimestampReceivedSecondsFraction = binary.BigEndian.Uint32(data[28:32])
+
+	currentTLVStartOffset := 32
+	for currentTLVStartOffset < len(data) {
+		currentTLV := &MPLSEchoTLV{}
+		numBytesDecoded, err := currentTLV.DecodeFromBytes(data[currentTLVStartOffset:])
+		if err != nil {
+			return err
+		}
+
+		m.TLVs = append(m.TLVs, currentTLV)
+		currentTLVStartOffset += numBytesDecoded
+	}
+
+	if currentTLVStartOffset != len(data) {
+		return fmt.Errorf(
+			"Invalid MPLSEcho content - TLVs don't match the valid data bounds (Total Valid Bytes = %v, Last TLV End Offset = %v)",
+			len(data),
+			currentTLVStartOffset)
+	}
+
+	return nil
+
+}
+
+// SerializeTo writes the serialized form of this layer into the SerializationBuffer, implementing gopacket.SerializableLayer.
+// See the docs for gopacket.SerializableLayer for more info.
+func (m *MPLSEcho) SerializeTo(b gopacket.SerializeBuffer, opts gopacket.SerializeOptions) error {
+	totalBytes := 32
+	for _, currentTLV := range m.TLVs {
+		valueByteCount := uint(len(currentTLV.Value))
+		tlvByteCountWithPadding := 4 + roundUpToNearestMultiple(valueByteCount, 4) // Pad to a multiple of 4 bytes
+		totalBytes += int(tlvByteCountWithPadding)
+	}
+
+	bytes, err := b.PrependBytes(totalBytes)
+	if err != nil {
+		return err
+	}
+
+	binary.BigEndian.PutUint16(bytes[0:], uint16(m.VersionNumber))
+	binary.BigEndian.PutUint16(bytes[2:], uint16(m.GlobalFlags))
+	bytes[4] = uint8(m.MessageType)
+	bytes[5] = uint8(m.ReplyMode)
+	bytes[6] = uint8(m.ReturnCode)
+	bytes[7] = uint8(m.ReturnSubcode)
+	binary.BigEndian.PutUint32(bytes[8:], m.SenderHandle)
+	binary.BigEndian.PutUint32(bytes[12:], m.SequenceNumber)
+	binary.BigEndian.PutUint32(bytes[16:], m.TimestampSentSeconds)
+	binary.BigEndian.PutUint32(bytes[20:], m.TimestampSentSecondsFraction)
+	binary.BigEndian.PutUint32(bytes[24:], m.TimestampReceivedSeconds)
+	binary.BigEndian.PutUint32(bytes[28:], m.TimestampReceivedSecondsFraction)
+
+	currentTLVStartOffset := 32
+	for _, currentTLV := range m.TLVs {
+		tlvByteRepresentation, err := currentTLV.EncodeAsBytes()
```

Comment:
This allocates a byte slice for every TLV. Memory allocations should be avoided. Can you change `EncodeAsBytes` so it writes to the target buffer directly?